internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -922,13 +950,34 @@ func (r *AgentServiceConfigReconciler) newAssistedServiceDeployment(log logrus.F
 		var replicas int32 = 1
 		deployment.Spec.Replicas = &replicas
 		deployment.Spec.Strategy = deploymentStrategy
+
+		// Handle our hashed configMap(s)
+		if deployment.Spec.Template.Annotations == nil {
+			deployment.Spec.Template.Annotations = make(map[string]string)
+		}
+		deployment.Spec.Template.Annotations[assistedConfigHashAnnotation] = assistedConfigHash
+		if mirrorConfigHash != "" {
+			deployment.Spec.Template.Annotations[mirrorConfigHashAnnotation] = mirrorConfigHash
+		}
+		if userConfigHash != "" {
+			runningUserConfigHash, ok := deployment.Spec.Template.Annotations[userConfigHashAnnotation]
+			if ok {
+				if runningUserConfigHash != userConfigHash {
+					msg := fmt.Sprintf("User provided and running configuration differ on assisted-service deployment. Remove the '%s' annotation from deployment to force restart", userConfigHashAnnotation)
```

Comment:
The error message confuses me a bit... If user CM and running user CMs do not match shouldn't we propose user to just restart/redeploy so that his changes are applied ?

Reply:
This feels like a downside to this approach. Because in order to detect a difference between the specified config and what the assisted-service is running, we need something like this hash. But now the only way to resolve it is to either:
1) Delete the assisted-service deployment, allowing the controller to recreate it with the correct annotations
2) Remove the specific hash annotation from the assisted-service deployment, the controller then can update the deployment

Reply:
I now see a [bug in how I wrote this section](https://github.com/openshift/assisted-service/pull/2137/files#r661585822) but the suggestion I linked is how I envision it working in my comment :arrow_up: